### DISCUSSION TITLE
Add the anthropic api interface

### DISCUSSION THREAD

**Discussion by 395273034** [open-webui/open-webui#1253] [General]
Please add the anthropic api interface, there is indeed a need for use, thank you!

https://openwebui.com/f/justinrahb/anthropic

**Comment by justinh-rahb** [comment]
Supported via LiteLLM integration:
- #1038

**Reply by bderubinat** [reply]
Not anymore. So, how do we add the anthropic api interface, mistral api, etc

**Reply by yuchenxie4645** [reply]
Deprecated after v0.2.4

**Reply by justinh-rahb** [reply]
Anthropic, Cohere, Azure, Groq (native API): https://github.com/open-webui/pipelines

Mistral, Groq, OpenRouter and many more have standard OpenAI-compatible APIs and you always could and still can add them as "OpenAI" endpoints in our **Settings > Connections** window.

**Reply by CaoHoangTung** [reply]
I'm trying to add Anthropic with the url `https://api.anthropic.com/v1`, seems like it's not working and not showing the list of models in the main chat.

**Reply by zaptrem** [reply]
I think anthropic is missing the /models endpoint open-webui expects.

**Reply by Seviks** [reply]
You can add a pipeline:

1. **Run [pipelines](https://github.com/open-webui/pipelines)** in Docker container
Run  `docker run -d -p 9099:9099 --add-host=host.docker.internal:host-gateway -v pipelines:/app/pipelines --name pipelines --restart always ghcr.io/open-webui/pipelines:main`

2. In OpenWebui go to **Admin Panel -> Connections -> OpenAI API**. Click on the "plus" sign. 
    Set the API URL to `http://localhost:9099` and the API key to `0p3n-w3bu!`. Or `host.docker.internal:9099` if localhost ist not working.

3. Go to **Admin Panel -> Pipelines -> Install from Github URL**. Add: `https://github.com/open-webui/pipelines/blob/main/examples/pipelines/providers/anthropic_manifold_pipeline.py` Add only trusted sources!

4. Add the API Key
 
I added pipelines in a docker compose file:
```
  pipelines:
    container_name: pipelines
    image: ghcr.io/open-webui/pipelines:main
    depends_on:
      - open-webui
    restart: unless-stopped
    volumes:
      - pipelines:/app/pipelines
```
In this case, use ``http://CONATAINER_NAME:9099`

**Reply by SchneiderSam** [reply]
> You can add a pipeline:

Would be nice if we get a cleaner OpenAI-API-Way

**Reply by DataCraftsmanAU** [reply]
Thanks, this pipeline method worked well.

**Reply by SchneiderSam** [reply]
Ok, now I'm a bit confused. After the newest Update

Shouldn't the new functions be used for this? I still find it irritating that there is an option to create connections in the admin area. But only to OpenAI API. But not to others. We have connections and pipelines for that. I don't think it's very clearly structured. What is the most stable and secure way?

**Reply by justinh-rahb** [reply]
> Ok, now I'm a bit confused. After the newest Update
> 
> Shouldn't the new functions be used for this? I still find it irritating that there is an option to create connections in the admin area. But only to OpenAI API. But not to others. We have connections and pipelines for that. I don't think it's very clearly structured. What is the most stable and secure way?

I've put significant work into both the Pipelines and Functions implementations of Anthropic provider, this one (Functions) is the easiest IMO:
https://openwebui.com/f/justinrahb/anthropic

**Reply by realgeneraliroh** [reply]
> Ok, now I'm a bit confused. After the newest Update
> 
> Shouldn't the new functions be used for this? I still find it irritating that there is an option to create connections in the admin area. But only to OpenAI API. But not to others. We have connections and pipelines for that. I don't think it's very clearly structured. What is the most stable and secure way?

To touch on this further, every API has a slightly different way of being interacted with.  And every API needs a custom interaction framework made for it.

**Reply by zuli12-dev** [reply]
> > Ok, now I'm a bit confused. After the newest Update
> > Shouldn't the new functions be used for this? I still find it irritating that there is an option to create connections in the admin area. But only to OpenAI API. But not to others. We have connections and pipelines for that. I don't think it's very clearly structured. What is the most stable and secure way?
> 
> I've put significant work into both the Pipelines and Functions implementations of Anthropic provider, this one (Functions) is the easiest IMO: https://openwebui.com/f/justinrahb/anthropic

This works great, maybe if someone needs it, add the `anthropic-beta: max-tokens-3-5-sonnet-2024-07-15` which allows 8192 tokens then...

The Log would tell so:

> Exception: HTTP Error 400: {"type":"error","error":{"type":"invalid_request_error","message":"max_tokens: 8192 > 4096, which is the maximum allowed number of output tokens for claude-3-5-sonnet-20240620. To access up to 8192 output tokens, specify the 'anthropic-beta: max-tokens-3-5-sonnet-2024-07-15' header."}}

**Reply by coachlou** [reply]
I just use openrouter for all the models - openai or others. simple.

**Comment by 395273034** [comment]
Thanks，I can use it！

model name:
1. claude-3-haiku-20240307
2. claude-3-sonnet-20240229
3. claude-3-opus-20240229

model addres:
https://api.anthropic.com/v1/messages

**Comment by justinh-rahb** [comment]
> model addres: https://api.anthropic.com/v1/messages

There's no need to explicitly set the API URL. LiteLLM already uses the appropriate default.

**Comment by zuli12-dev** [comment]
i guess the problem is at the moment that there is no GET available for the Models which are offered:

> 2024-08-13 15:14:49 INFO  [apps.openai.main] get_all_models()
> 2024-08-13 15:14:49 INFO  [config] Saving 'OPENAI_API_BASE_URLS' to config.json
> 2024-08-13 15:14:49 INFO:     192.168.65.1:38487 - "POST /openai/urls/update HTTP/1.1" 200 OK
> 2024-08-13 15:14:49 INFO  [config] Saving 'OPENAI_API_KEYS' to config.json
> 2024-08-13 15:14:49 INFO:     192.168.65.1:38487 - "POST /openai/keys/update HTTP/1.1" 200 OK
> 2024-08-13 15:14:50 ERROR [apps.openai.main] 404 Client Error: Not Found for url: https://api.anthropic.com/v1/models
> 2024-08-13 15:14:50 Traceback (most recent call last):
> 2024-08-13 15:14:50   File "/app/backend/apps/openai/main.py", line 324, in get_models
> 2024-08-13 15:14:50     r.raise_for_status()
> 2024-08-13 15:14:50   File "/usr/local/lib/python3.11/site-packages/requests/models.py", line 1024, in raise_for_status
> 2024-08-13 15:14:50     raise HTTPError(http_error_msg, response=self)
> 2024-08-13 15:14:50 requests.exceptions.HTTPError: 404 Client Error: Not Found for url: https://api.anthropic.com/v1/models
> 2024-08-13 15:14:50 INFO:     192.168.65.1:38487 - "GET /openai/models/3 HTTP/1.1" 500 Internal Server Error
> 2024-08-13 15:14:50 INFO  [apps.openai.main] get_all_models()
> 2024-08-13 15:14:50 INFO  [apps.ollama.main] get_all_models()
> 2024-08-13 15:14:50 INFO:     192.168.65.1:38487 - "GET /api/models HTTP/1.1" 200 OK



Other's like groq, mistral, together.ai all have this available and thus work at the moment perfectly fine...

Would love if anthropic add's support for this, otherwise we will need to stick to a pipe or function...

**Comment by rcdailey** [comment]
If LiteLLM has been removed from Open WebUI, is this doc page relevant anymore? https://docs.openwebui.com/tutorial/litellm/

What exactly do I need to do to set this up? Unlike other areas of Open WebUI, this is not straightforward.

**Reply by rampageservices** [reply]
Curious to hear an update here from those in the know.

**Comment by mynumu** [comment]
Hello - wow a lots changed in the 5 months since I used this! looking very cool but er any reason we cant just add in litellm and openrouter as direct aPI keys like openai? Same for all the other LLM models - looks like its some complex pipeline setup??

Request - Add Anthropic API Version

It seems openrouter api direct connection works?

**Comment by iammrbt** [comment]
Yes, please add native anthropic support

**Comment by koiker** [comment]
I believe the only missing part to make Anthropic compatible with the OpenAI API is to include a feature in Open-WebUI interface to include extra HTTP headers in the Connections.
If you include a new extra field to add header key/value and allow the admin to input this value. We can add the header anthropic-version: 2023-06-01
This will make the API calls to Anthropic compatible to list models, and all other API calls should work as well.

**Comment by sitzbrau** [comment]
if i import the [Anthropic functions](https://openwebui.com/f/justinrahb/anthropic) on my openwebui:ollama container, on a google cloud vm instance i've got this error
![image](https://github.com/user-attachments/assets/d65c56b2-9fda-4c16-9042-89988b731358)

**Reply by marleyjaffe** [reply]
I get this same error

**Comment by giadefa** [comment]
Antrophic does not provide /models, so you need to have them explicitly in the models part of the webui form and then it works out of the box, e.g. use the url https://api.anthropic.com/v1, put the API key and add the explict name of the model and it works

**Reply by corticalstack** [reply]
> Antrophic does not provide /models, so you need to have them explicitly in the models part of the webui form and then it works out of the box, e.g. use the url https://api.anthropic.com/v1, put the API key and add the explict name of the model and it works

Unfortunately I couldn't get this working as described, with openwebui v0.6.15

**Reply by koiker** [reply]
> Antrophic does not provide /models, so you need to have them explicitly in the models part of the webui form and then it works out of the box, e.g. use the url https://api.anthropic.com/v1, put the API key and add the explict name of the model and it works

Anthropic does have a /models API. [Doc here](https://docs.anthropic.com/en/api/models-list)
The problem is the required headers to make this API call is not included in the standard OpenAPI (In other words, Anthropic have a similar but not compliant REST API with OpenAI)

I believe the suggestion in this discussion is to let the Open WEBUI have a customizable headers parameter where you can add extra HTTP headers to communicate with some API endpoints.
In this case, we can add the required HTTP header that Anthropic requires to make the /v1/models work.
Also, this feature could help other integrations that may require extra headers in the API request.

**Reply by magnetic5355** [reply]
@giadefa  - what is the format of the model name?

**Comment by giadefa** [comment]
Whatever model you would like to use, e.g. claude-sonnet-4-20250514

**Comment by Alexanderamiri** [comment]
Openwebui should have support for this natively. I find it limiting that it's only openai in and out. This project could've really been a go to if it supported more providers out of the box and perhaps even schema translation like litellm does

**Comment by Podden** [comment]
Just use this 🫡 
https://openwebui.com/f/podden/anthropic_pipe

**Comment by esauvisky** [comment]
I figured out what's causing this, and I opened an [issue](https://github.com/open-webui/open-webui/issues/18929) with a suggested solution.

**Comment by xylphe** [comment]
I am discovering OpenWebUI and am unable to use Anthropic models.

From what I have read, I see that there are several options:
* OpenRouter
* LiteLLM
* Using a community function.

Do you know which option offers the best integration of Claude's features with OpenWebUI?

**Reply by Classic298** [reply]
depends. the quickest way is installing a community made pipe

the best ways in terms of compatibility are openrouter and litellm

**Reply by Alexanderamiri** [reply]
its silly to only support openai natively and demote all other providers to be used through pipes. This project should've had modular design in providers and had support for MCP.

**Reply by Classic298** [reply]
@Alexanderamiri why should Open WebUI with very limited development resources and very limited active contributors, VOLUNTEER CONTRIBUTORS WITH FULL TIME JOBS, waste away it's time to implement more than a dozen different API schemas?

Open WebUI only implements widely adopted API standards and this has been the case for years and is clearly written in the docs.

Most providers support OpenAI API. The vast majority in fact.

Anthropic API is only Anthropic.
Why should we waste OUR time to support one single provider who does not want to adopt a widely spread API standard?

**Reply by Classic298** [reply]
https://docs.openwebui.com/faq#q-why-doesnt-open-webui-support-specific-providers-latest-api-eg-openai-responses-api


https://docs.openwebui.com/getting-started/quick-start/starting-with-openai#important-protocols-not-providers

**Reply by Classic298** [reply]
PS: the pipes are exactly this modular design you describe

One-Click install a community pipe and boom you got full anthropic API support

**Reply by Classic298** [reply]
Check out the community on openwebui.com there are many more pipes, many provider specific pipes, api translation... even features, plugins, tools, filters... and a lot more.

**Reply by Alexanderamiri** [reply]
@Classic298 It's not waste of time to support more providers at least not the biggest ones eg bedrock, vertex, azure, openai, anthropic where most people and companies get their inference. There has also been contributors trying to add support for more providers but the clear message being "just use pipes". 

This project also doubled down on openapi instead of mcp a while ago and communicated that openapi is a more neutral standard. While all other llm related project support mcp as it has become the standard llm tooling spec. 

The project direction is lacking because of philosophical stands such as these. Code is cheap these days and adding a anthropic.py with a few houndred lines or and a few edits around the codebase similar to how there already is two providers, openai.py and ollama.py, doesnt strike me as wasted time or effort if the time spent brings value to a portion of the users. More mental capacity is used on responding to people who want these features and the constant reminding of different prs that we will not add mcp native or more providers.

I would contribute if i knew it wouldn't be wasted and ignored but countless discussions here shows that the issue isn't limited time on not enough interest in these features but rather a stubborn take on standards. I maintain a openwebui fork used by 500 people with most of these changes implemented and it takes me little effort to fix bugs when either anthropic, openai, claude code or any other tool or provider updates their fragile services. Like I said code is cheap.

Ref discussion below where there is a lot of interest on a simple implementation of mcp native
https://github.com/open-webui/open-webui/discussions/16238#discussion-8667954

**Reply by Classic298** [reply]
<ins>**The project supports MCP.**</ins> I think you are very outdated here.

you said you wanted modularity - pipes are modularity.

And companies that require enterprise vertex AI or Azure typically use LiteLLM or other middlewars for other reasons anyways. Tracking, api keys, model management, tracability and logging and more.

and adding anthropic.py is not something that is cheap and can be done with a few hundred lines easily.
Take a look at the openai.py. More than a thousand lines
**And also take a look at the frontend.**

Reasoning content by anthropic API is rendered entirely different. Requires changes in vast amounts of the codebase. Requires flaky changes on the frontend to detect the different reasoning types, likely breaking existing behaviour.

Tool calling is different - which requires, again, changes in many parts of the backend code but also frontend to handle the tool calls correctly.
Anthropic API standard is different in many ways to the OpenAI standard - and the frontend and the backend are fully trimmed to that.

> It's not waste of time to support more providers at least not the biggest ones eg bedrock, vertex, azure, openai, anthropic

Now you are going to say, once Anthropic API is implemented

"great, now that we got that, why not vertex, why not bedrock, why not Google's new interactions API"

**boom - pandora's box is open**

Because you will flood the codebase with nonsense.
You will need different handling for all the different api schemas in every single file.
You will need different frontend handling for every single API

This is NEITHER trivial, NOR cheap code, NOR without risk since you are definitely at risk of breaking other implementations.
"Code is cheap" is easy to say when you can live with minor bugs here and there. But when you are working on a 123k star project with 200 mil downloads used by individuals and large corporations alike - you need stability and cannot afford breakage. A 500 user deployment? So what if it's down for 1hr - let me quickly hotfix the Anthropic integration - and done.

**Reply by Classic298** [reply]
This is less of a philosophical stance but, once again, a stance of

- economic feasibility - it does not make sense to spend time on this when you can solve it with pipes (modularity) or middlewares easily
- **volunteer** contributor time
- and it's not as trivial as you think to make a stable, production ready integration of this without breaking ANYTHING of the existing behaviour, neither native tool calling, tool rendering in the UI, the thinking blocks etc.
- and of course this would require a lot of testing
- and finally, this would open the pandora's box - for anyone to ask for provider specific integrations, which will cost even more volunteer contributor time and make it even harder to maintain long term

And the users who demanded the integration will be the first to complain when it breaks, even slightly, in a new version.

Plus of course the time spent maintaining all the different integrations, all the different API schemas down the line, when one slightly changes, it can be devastating to the whole integration requiring vast changes in the frontend and backend codebase.

Maintaining it is the real burden.
The more integrations - the more complex it becomes.

> More mental capacity is used on responding to people

I know what you want to say here.

But people also need to understand the perspective of volunteer contributors like me.

In fact, yes, it is easier to respond here, than write and later on maintain the code.

If you think it is easier to write and maintain the code down the line, feel free to become a volunteer contributor and ensure the integration stays intact forever. This is easily done when you only have your own small deployment. This is not easily done when you have a scenario in which you have 10 provider integrations, one updates their API, and suddenly your entire code is useless and you have to refactor all of it and still ensure compatibility with the code of the other integrated providers.

**Reply by Podden** [reply]
https://openwebui.com/search?query=Anthropic&sort=top&t=all

It's a little bit tricky to find "the best" one because sorting by Downloads or Top isn't working as expected and obviously I dont know what you need. I would propose that you try the more recently updated ones which have many downloads and/or upvotes and work your way through.

Generally I find those function are the best way to get the providers in. You can really focus on getting the best translation layer between openwebui and the quirks of the Individual API. Uploading a native pdf file (without RAG) for Anthropic Code Execution or using the files api or programmatic Tool calling is only possible through a function/pipe.

**Comment by tblattner** [comment]
Sorry forgot to search before creating my discussion. I have a working implementation that adds this end point as a proxy: https://github.com/open-webui/open-webui/discussions/21390

**Comment by Classic298** [comment]
In case anyone here was not aware, Anthropic has an OpenAI compatible endpoint, which works just fine for all of their models.

If you don't want to use a pipe, then you can just use their OpenAI compatible API and it works just fine.
It's even documented [here](https://docs.openwebui.com/getting-started/quick-start/connect-a-provider/starting-with-openai-compatible)

I see less and less reasons why a native integration is needed. Open WebUI is modular and supports pipes for this exact reason - to implement any provider you want. Middlewares exist and Anthropic even offers their own OpenAI compatible endpoint.

**Reply by realgeneraliroh** [reply]
This is wonderful news to me!

**Reply by Podden** [reply]
Small addendum: be aware the anthropics openai compatibility layer does not Support some Important Features, the most missed one being promt caching. For multi turn conversations like openwebui provides, this will really up your spending unnecessary.

**Reply by Classic298** [reply]
does anthropic support implicit caching? I have never seen it cache any prompts.

Only explicit caching exists if i am not mistaken?

**Reply by Podden** [reply]
Yes, only explizit caching! I just wanted to say that this would be an Important factor to think about when deciding about how to access their models.

**Reply by Classic298** [reply]
In that case, i am not sure explicit caching is of much use in open webui. Hard to make use of it

**Reply by Podden** [reply]
Still, I put lots of effort in my pipe, sorting and extracting and moving around, to preserve the Cache while supporting all openwebui features that would otherwise alter message history and system promt like, file uploads, memory and RAG. Saved mucho dineros with that 😉.
